msal/tests/Test.MSAL.NET.Unit/ConfidentialClientApplicationTests.cs

```diff
@@ -295,392 +314,514 @@ private ConfidentialClientApplication CreateConfidentialClient(ClientCredential
-            Assert.AreEqual(TestConstants.ClientId, qp["client_id"]);
-            Assert.AreEqual("code", qp["response_type"]);
-            Assert.AreEqual(TestConstants.RedirectUri, qp["redirect_uri"]);
-            Assert.AreEqual(TestConstants.DisplayableId, qp["login_hint"]);
-            Assert.AreEqual(UIBehavior.SelectAccount.PromptValue, qp["prompt"]);
-            Assert.AreEqual("MSAL.Desktop", qp["x-client-sku"]);
-            Assert.IsFalse(string.IsNullOrEmpty(qp["x-client-ver"]));
-            Assert.IsFalse(string.IsNullOrEmpty(qp["x-client-cpu"]));
-            Assert.IsFalse(string.IsNullOrEmpty(qp["x-client-os"]));
-
-            Assert.IsTrue(HttpMessageHandlerFactory.IsMocksQueueEmpty, "All mocks should have been consumed");
+                httpManager.AddInstanceDiscoveryMockHandler();
+
+                ConfidentialClientApplication app = new ConfidentialClientApplication(
+                    httpManager,
+                    TestConstants.ClientId,
+                    ClientApplicationBase.DefaultAuthority,
+                    TestConstants.RedirectUri,
+                    new ClientCredential(TestConstants.ClientSecret),
+                    new TokenCache(),
+                    new TokenCache())
+                {
+                    ValidateAuthority = false
+                };
+
+                //add mock response for tenant endpoint discovery
+                httpManager.AddMockHandler(
+                    new MockHttpMessageHandler
+                    {
+                        Method = HttpMethod.Get,
+                        ResponseMessage = MockHelpers.CreateOpenIdConfigurationResponse(app.Authority)
+                    });
+
+                Task<Uri> task = app.GetAuthorizationRequestUrlAsync(TestConstants.Scope, TestConstants.DisplayableId, null);
+                Uri uri = task.Result;
+                Assert.IsNotNull(uri);
+                Dictionary<string, string> qp = CoreHelpers.ParseKeyValueList(uri.Query.Substring(1), '&', true, null);
+                Assert.IsNotNull(qp);
+                Assert.AreEqual(11, qp.Count);
+                Assert.IsTrue(qp.ContainsKey("client-request-id"));
+                Assert.AreEqual("offline_access openid profile r1/scope1 r1/scope2", qp["scope"]);
+                Assert.AreEqual(TestConstants.ClientId, qp["client_id"]);
+                Assert.AreEqual("code", qp["response_type"]);
+                Assert.AreEqual(TestConstants.RedirectUri, qp["redirect_uri"]);
+                Assert.AreEqual(TestConstants.DisplayableId, qp["login_hint"]);
+                Assert.AreEqual(UIBehavior.SelectAccount.PromptValue, qp["prompt"]);
+                Assert.AreEqual("MSAL.Desktop", qp["x-client-sku"]);
+                Assert.IsFalse(string.IsNullOrEmpty(qp["x-client-ver"]));
+                Assert.IsFalse(string.IsNullOrEmpty(qp["x-client-cpu"]));
+                Assert.IsFalse(string.IsNullOrEmpty(qp["x-client-os"]));
```

Comment:
seems like most of the validation stuff is the same in more places. ... seems duplicate and makes me think that some refactorings are needed and perhaps the tests are not testing different things...

Reply:
yes, i had similar feelings on copy/paste and duplication across many of the tests, this being a prime example.  test cleanup and organization is also in my backlog.